packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java

```diff
@@ -375,17 +376,54 @@ public ResponseEntity<StreamingResponseBody> getFile(
 	public ResponseEntity<PresignedURL> getDownloadURL(
 			@PathVariable("id") final UUID id,
 			@RequestParam("filename") final String filename) {
-
+		final Optional<Dataset> dataset;
 		try {
-			final Optional<PresignedURL> url = datasetService.getDownloadUrl(id, filename);
-			return url.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
-		} catch (final Exception e) {
-			final String error = "Unable to get download url";
+			dataset = datasetService.getAsset(id);
+			if (dataset.isEmpty()) {
+				throw new ResponseStatusException(
+					org.springframework.http.HttpStatus.NOT_FOUND,
+					"Dataset not found");
+			}
+		} catch (final IOException e) {
+			final String error = "Unable to get dataset";
 			log.error(error, e);
 			throw new ResponseStatusException(
 					org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR,
 					error);
 		}
+
+		if(dataset.get().getEsgfId() != null && !dataset.get().getEsgfId().isEmpty()
+			&& dataset.get().getDatasetUrls() != null && !dataset.get().getDatasetUrls().isEmpty()){
```

Comment:
Liking - super solid

Reply:
In general I prefer java to javascript... but I do love the brevity of just checking `if(!someVar)` and that does a null and empty check at the same time...